File: app/src/main/java/com/dmko/pairwisecomparison/ui/screens/addeditcomparison/AddEditComparisonPresenter.java
package com.dmko.pairwisecomparison.ui.screens.addeditcomparison;

import com.dmko.pairwisecomparison.data.entities.Comparison;
import com.dmko.pairwisecomparison.data.repositories.ComparisonRepository;
import com.dmko.pairwisecomparison.ui.base.mvp.impl.BasePresenterImpl;
import com.dmko.pairwisecomparison.utils.SchedulersFacade;

import timber.log.Timber;

import static com.dmko.pairwisecomparison.utils.LogTags.LOG_APP;

public class AddEditComparisonPresenter extends BasePresenterImpl<AddEditComparisonContract.View> implements AddEditComparisonContract.Presenter {

    private SchedulersFacade schedulers;
    private ComparisonRepository repository;

    private String comparisonId;
    private Comparison comparison;

    public AddEditComparisonPresenter(SchedulersFacade schedulers, ComparisonRepository repository) {
        this.schedulers = schedulers;
        this.repository = repository;
    }

    @Override
    public void setArgs(String comparisonId) {
        this.comparisonId = comparisonId;
    }

    @Override
    public void start() {
        Timber.tag(LOG_APP);
        Timber.i("Starting %s with comparisonId = %s", this.getClass().getSimpleName(), comparisonId);

        getView().showLoading(true);
        if (comparisonId == null) {
            comparison = new Comparison();
            getView().showLoading(false);
        } else {
            addDisposable(repository.getComparisonById(comparisonId)
                    .subscribeOn(schedulers.io())
                    .observeOn(schedulers.ui())
                    .subscribe(comparison -> {
                        Timber.tag(LOG_APP);
                        Timber.i("%s sending to %s, %s", this.getClass().getSimpleName(), getView().getClass().getSimpleName(), comparison);

                        if (isViewAttached()) {
                            this.comparison = comparison;
                            getView().setComparison(comparison);
                            getView().showLoading(false);
                        }
                    }));
        }
    }

    @Override
    public void saveComparison(String newName) {
        if (newName.isEmpty()) {
            getView().showEmptyNameError();
        } else {
            comparison.setName(newName);
            if (comparisonId == null) {
                addDisposable(repository.insertComparison(comparison)
                        .subscribeOn(schedulers.io())
                        .subscribe(() -> {
                            if (isViewAttached()) {
                                getView().closeDialog();
                            }
                        }));
            } else {
                addDisposable(repository.updateComparison(comparison)
                        .subscribeOn(schedulers.io())
                        .subscribe(() -> {
                            if (isViewAttached()) {
                                getView().closeDialog();
                            }
                        }));
            }
        }
    }
}

File: app/src/main/java/com/dmko/pairwisecomparison/ui/screens/addeditcomparison/AddEditComparisonPresenter.kt
package com.dmko.pairwisecomparison.ui.screens.addeditcomparison

import com.dmko.pairwisecomparison.data.entities.Comparison
import com.dmko.pairwisecomparison.data.repositories.ComparisonRepository
import com.dmko.pairwisecomparison.ui.base.mvp.impl.BasePresenterImpl
import com.dmko.pairwisecomparison.utils.LogTags.LOG_APP
import com.dmko.pairwisecomparison.utils.SchedulersFacade
import timber.log.Timber

class AddEditComparisonPresenter(private val schedulers: SchedulersFacade, private val repository: ComparisonRepository) : BasePresenterImpl<AddEditComparisonContract.View>(), AddEditComparisonContract.Presenter {

    private var comparisonId: String? = null
    private lateinit var comparison: Comparison

    override fun setArgs(comparisonId: String?) {
        this.comparisonId = comparisonId
    }

    override fun start() {
        Timber.tag(LOG_APP)
        Timber.i("Starting %s with comparisonId = %s", this.javaClass.simpleName, comparisonId)

        view?.showLoading(true)
        if (comparisonId == null) {
            comparison = Comparison()
            view?.showLoading(false)
        } else {
            addDisposable(repository.getComparisonById(comparisonId)
                    .subscribeOn(schedulers.io())
                    .observeOn(schedulers.ui())
                    .subscribe { comparison ->
                        Timber.tag(LOG_APP)
                        Timber.i("%s sending to %s, %s", this.javaClass.simpleName, view.javaClass.simpleName, comparison)

                        this.comparison = comparison
                        view?.setComparison(comparison)
                        view?.showLoading(false)
                    })
        }
    }

    override fun saveComparison(comparisonName: String) {
        if (comparisonName.isEmpty()) {
            view?.showEmptyNameError()
        } else {
            comparison.name = comparisonName
            if (comparisonId == null) {
                addDisposable(repository.insertComparison(comparison)
                        .subscribeOn(schedulers.io())
                        .subscribe {
                            view?.closeDialog()
                        })
            } else {
                addDisposable(repository.updateComparison(comparison)
                        .subscribeOn(schedulers.io())
                        .subscribe {
                            view?.closeDialog()
                        })
            }
        }
    }
}
